### PR TITLE
dynamic imports fix  for non-webpack environments.

### PR DESCRIPTION
**Pull request by PrudviGali**
Added issue description in #3345 .

Feature detection condition comes handy rather than `typeof window` because of webpack specific methods `require.resolveWeak` and `resolve.ensure` fails in non-webpack environments.

The issue was not with `window` object but with `require.resolveWeak` and `require.ensure` methods used in next handle-import.js file. Both these methods are specific to webpack and would not be available in `node/commonjs` environment.  Jest runs in Node environment and does not recognize `require.resolveWeak` and `require.ensure` methods.

TC39 [dynamic import proposal](https://github.com/tc39/proposal-dynamic-import) works pretty well with Jest testing. Please refer to the [repo](https://github.com/PrudviGali/dynamic-imports-jest-testing). But Nextjs implementation of dynamic imports fails in jest because of above two methods [repo](https://github.com/PrudviGali/nextjs-dynamic-import-test-fail).

@arunoda already mentioned that should not use `process.env.NODE_ENV === 'test'` in #3405. Assuming this would be the appropriate way of handling this issue #3345 .

### CODE REVIEW

**Comment by arunoda** [comment]
Oh! I didn't see this: https://github.com/zeit/next.js/pull/3414
(Was thinking it's something else)

Sorry for that.
Looks interesting.

File: server/build/babel/plugins/handle-import.js [hunk 1]
```diff
@@ -9,7 +9,7 @@ const TYPE_IMPORT = 'Import'
 
 const buildImport = (args) => (template(`
   (
-    typeof window === 'undefined' ?
+    typeof require.resolveWeak !== 'function' ?
```
Comment (arunoda) [inline server/build/babel/plugins/handle-import.js]: Could you add a comment where why we use this rather a simple `window` check.
It's better for someone who's reading the code.

**Review by arunoda — APPROVED** [review]

**Review by arunoda — CHANGES_REQUESTED** [review]
Changing the status.

**Comment by PrudviGali** [comment]
@arunoda No problem. I should have added documentation in the first place.

Anyhow added the requested changes.

**Review by PrudviGali — COMMENTED** [review]
Added requested changes

**Review by arunoda — APPROVED** [review]

**Comment by arunoda** [comment]
@PrudviGali this looks great.
Thanks.

**Comment by PrudviGali** [comment]
@arunoda Cool. Thanks.

**Comment by timneutkens** [comment]
Very nice, thanks @PrudviGali!

**Pull request merged**